klic-3.014/documents/workshop/tgif/tgif-QPL-4.2.5/tgif.man
.\" Tgif's man pages.
.\"
.\" @(#)$Header: /mm2/home/cvs/bc-src/tgif/tgif.man,v 1.104 2011/05/23 02:56:57 william Exp $
.\"
.\"
.\"
.\"
.\"
.\"
.\"
.TH tgif n "Version 4.2 Patchlevel 3 and Above" "Tgif"
.\"
.SH NAME
.\"
\fItgif\fR \- Xlib based interactive 2-D drawing facility under X11.
Supports hierarchical construction of drawings and easy navigation between
sets of drawings.  It's also a hyper-graphics
(or hyper-structured-graphics) browser on the World-Wide-Web.
.\"
.SH SYNOPSIS
.\"
.B tgif
[\fB-display\fR displayname]
[\fB-fg\fR <color>]
[\fB-bg\fR <color>]
[\fB-bd\fR <color>]
[\fB-rv\fR]
[\fB-nv\fR]
[\fB-bw\fR]
[\fB-reqcolor\fR]
[\fB-cwo[+sbwarp]\fR]
[\fB-hyper\fR]
[\fB-exec <file>\fR]
[\fB-dbim {xcin|chinput|xim|kinput2|tgtwb5[,font]}\fR]
[\fB-sbim xim\fR]
[\fB-usexlib\fR]
[\fB{-a4|-letter}\fR]
[\fB-listdontreencode\fR]
[\fB-version\fR]
[\fB-pdfspd | -pdfspd=true | -pdfspd=false \fR]
[\fB-pssetup "<string>" \fR]
[\fB-tgwb2 [-rmcastlibdir <directory> | -rmcastlibpath <path>]\fR]
[\fB-nomode\fR]
[\fB-geometry\fR <geom>]
[\fB=<geom>\fR]
[{\fIfile\fR[.\fIobj\fR]|\fB-merge\fR
\fIfile1\fR[.\fIobj\fR] \fIfile2\fR[.\fIobj\fR] ...}]
.PP
or
.PP
.B tgif
\fB-print\fR
[\fB-eps\fR]
[\fB-p\fR]
[\fB-ps\fR]
[\fB-f\fR]
[\fB-text\fR]
[\fB-epsi\fR]
[\fB-tiffepsi\fR]
[\fB-gif\fR]
[\fB-png\fR]
[\fB-jpeg\fR]
[\fB-ppm\fR]
[\fB-pbm\fR]
[\fB-xpm\fR]
[\fB-xbm\fR]
[\fB-html\fR]
[\fB-pdf\fR]
[\fB-netlist\fR]
[\fB-svg\fR]
[\fB-display\fR displayname]
[\fB-stdout\fR]
[\fB-raw[+h[eaderonly]]\fR]
[\fB-dosepsfilter\fR [\fB-previewonly\fR]]
[\fB-status\fR]
[\fB-gray\fR]
[\fB-color | -reqcolor\fR]
[\fB-adobe | -adobe=<number>/<number> | -adobe=false \fR]
[\fB-dontreencode=<string> | -listdontreencode\fR]
[\fB-version\fR | \fB-justversion\fR]
[\fB-producedby=<string>\fR]
[\fB-page <number>\fR]
[\fB-print_cmd "<command>"\fR]
[\fB-one_file_per_page\fR]
[\fB-pepsc\fR]
[\fB-pdfspd | -pdfspd=true | -pdfspd=false \fR]
[\fB-pssetup "<string>" \fR]
[\fB-j2p6_cmd "<command>" \fR]
[\fB-dontcondense | -condensed\fR]
[\fB{-a4|-letter}\fR]
[\fB-noshowpageineps\fR]
[\fB-quiet\fR]
[\fB-bop_hook "<string>"\fR]
[\fB-eop_hook "<string>"\fR]
[\fB-tmp_file_mode "<octal number>"\fR]
[\fB-patterndir "<xbm directory>"\fR]
[\fB-o\fR<dir>]
[\fB-exec <file>\fR]
[\fIfile1\fR[.\fIobj\fR] \fIfile2\fR[.\fIobj\fR] ...]
.\"
.SH DESCRIPTION
.\"
\fITgif\fR is an interactive drawing tool that allows the user to draw and
manipulate objects in the X Window System.  Tgif runs interactively in
the first form.
In the second form shown in the SYNOPSIS section,
tgif just prints \fIfile1.obj\fR, \fIfile2.obj\fR,
etc.  (generated by \fItgif\fR)
into PostScript(TM) page description files
(without opening windows or fonts)
and pipes them to lpr(1) if none of the
\fB-eps\fR, \fB-p\fR, \fB-epsi\fR, \fB-tiffepsi\fR, \fB-gif\fR,
\fB-png\fR, \fB-jpeg\fR, \fB-ppm\fR, \fB-pbm\fR, \fB-xpm\fR, \fB-xbm\fR,
\fB-html\fR, \fB-pdf\fR, \fB-ps\fR, \fB-f\fR, \fB-text\fR, \fB-netlist\fR, or
\fB-svg\fR options
are specified.  This form of \fIprinting\fR is tgif's way of \fIexporting\fR
a tgif file to another format.  In this case,
any other unrecognized command line options are sent to lpr(1).
In this mode, tgif is compatible with the obsoleted \fIprtgif\fR.
A symbol file (see descriptions below) can also be printed by
specifying the \fI.sym\fR extension explicitly.
.PP
The command line argument \fIfile\fR
specifies a file or an Uniform Resource Locator (URL) of
objects to be initially edited by tgif.
Only HTTP or FTP URL's are supported.
(For a more detailed description of URL and the World-Wide-Web,
the reader is referred to [1].)
.PP
Tgif is purely based on \fIXlib\fR.
It is tested under X11R6, and
it requires a 3 button mouse.
.\"
.SH OPTIONS
.\"
In the first form shown in the SYNOPSIS section,
the command line arguments can be:
.TP
.B -fg
Foreground color specified in <color>.
.TP
.B -bg
Background color specified in <color>.
.TP
.B -bd
Border color specified in <color>.
.TP
.B -rv
Start tgif in reversed-video mode.
.TP
.B -nv
Start tgif in normal-video mode.
.TP
.B -bw
Start tgif in black and white mode.
.TP
.B -reqcolor
Same effect as setting the
Tgif.PrintUsingRequestedColor X default to true
(see the X DEFAULTS section below).
.TP
.B -cwo
Canvas Window Only.  Only the canvas window (see TGIF SUBWINDOWS
section below) will be displayed.
This has the same effect as setting the Tgif.CanvasWindowOnly X default to true.
.TP
.B -cwo+sbwarp
If \fB-cwo+sbwarp\fR is used, single-button-warp (clicking the
left mouse button to warp) is used to activate teleporting
(see TELEPORT/HYPERJUMP section below).
.TP
.B -hyper
Start tgif in the \fIhyperspace\fR mode
(see HYPERSPACE section below).
.TP
.B -exec <file>
After tgif starts, execute the internal command in <file>
(see INTERNAL COMMANDS section below).
If <file> is the string "-", tgif executes internal commands from the
standard input.
.TP
.B -dbim \fImethod\fR
Use \fImethod\fR as the input method for double-byte fonts
(see SQUARE DOUBLE BYTE FONTS section below).
This cannot be used in conjunction with \fI-sbim\fR.
.TP
.B -sbim \fImethod\fR
Use \fImethod\fR as the input method for single-byte fonts.
This is useful if the X Keyboard Extension is used in
inputing international characters (with dead keys).
This cannot be used in conjunction with \fI-dbim\fR.
.TP
.B -usexlib
If tgif is compiled with -DUSE_XT_INITIALIZE, X Toolkit initialization
routines will be used to setup tgif.  Using this command line option will
force tgif to ignore the -DUSE_XT_INITIALIZE compiler option and use
Xlib only.  This is useful when the system resource file for tgif is
not installed properly or messed up and needs to be bypassed.
.TP
.B -a4
Using this option has the same effect as setting the Tgif.PSA4PaperSize
X default to true.
.TP
.B -letter
Using this option has the same effect as setting the Tgif.InitialPaperSize
X default to "letter"
.TP
.B -noshowpageineps
Using this option has the same effect as setting the Tgif.ShowPageInEPS
X default to false.
.TP
.B -quiet
If this option is used, tgif will suppress standard messages.
.TP
.B -listdontreencode=<string>
If this option is used, tgif will print out the list of PostScript
font names specified in the -D_DONT_REENCODE compiler option used
in compiling tgif.
.PP
In the second form shown in the SYNOPSIS section,
the command line arguments can be:
.TP
.B -version
If this option is used, tgif will print out its version number
and copyright on the command line.
.TP
.B -justversion

klic/ChangeLog
Thu Feb 13 17:11:01 2003  Yusuke KANEKI  <lang@ueda.info.waseda.ac.jp>

	* Configure, version.sed, Makefile.tail
	* runtime/Makefile.tail, Compiler/Makefile
	* runtime/gunix.kl1, runtime/asyncio.c
	support cygwin and gcc 3.xx .
	
Fri Jan 31 15:12:09 2003  Yusuke KANEKI  <lang@ueda.info.waseda.ac.jp>

	* include/klic/control.h, include/klic/gg_macro.h, runtime/debug.c
	fix the bug in control.h REDCTION_COUNT -> REDUCTION_COUNT
	fix the bug in gg_macro.h GG_ERROR_IN_IN -> GG_ERROR_IN_NEW
	fix the bug in debug.c sys_errlist[] -> strerror()
	
Thu Jan 30 14:44:50 2003  Ryo MATSUMURA  <lang@ueda.info.waseda.ac.jp>

	* runtime/alloc.c, runtime/gc.c, include/klic/struct.h:
	add functions for DKLIC.
	
Thu Jan 30 14:34:33 2003  Ryo MATSUMURA  <lang@ueda.info.waseda.ac.jp>

	* runtime/termio.kl1:
	fix the bug in full-duplex IO for socket stream.

Thu Jan 30 12:24:19 2003  Ryo MATSUMURA  <lang@ueda.info.waseda.ac.jp>

	* runtime/asyncio.c, runtime/gunix.kl1:
	support ip multi cast.

Sun Dec 28 18:56:12 2002  Yusuke KANEKI  <lang@ueda.info.waseda.ac.jp>

	* runtime/gunix.kl1:
	fix the bug in ungetc.

Sun Nov 18 12:22:34 2001  Xiobi WEI   <lang@ueda.info.waseda.ac.jp>

	* runtime/gunix.kl1:
	fix the bug in mktemp.
	
Thu Nov  1 20:07:47 2001  Yusuke KANEKI  <lang@ueda.info.waseda.ac.jp>

	* runtime/termio.kl1, runtime/gunix.kl1:
	fix the bug in fork_with_pipes.

Wed Oct 11 16:33:41 2000  Yusuke TAKAGI  <takagi@ueda.info.waseda.ac.jp>

	* runtime/termio.kl1: support prolog-like IO for socket stream.

Wed Oct 11 16:31:13 2000  Hiroshi IGARASHI  <iga@ruby-lang.org>

	* runtime/asyncio.c, runtime/gunix.kl1, runtime/signal.c:
	support full-duplex IO for socket stream.

Thu May  3 23:06:39 2001  Hiroshi IGARASHI  <iga@ruby-lang.org>

	* Makefile.tail, compiler/Makefile: specify -L../runtime via
	LD_LIBRARY_OPTIONS when compiler is compiled.

Thu Oct 12 18:16:32 2000  Hiroshi IGARASHI  <iga@ruby-lang.org>

	* Configure, Makefile.tail: specify runpath by ld option `-R'
	for Solaris.

Thu Oct 12 14:54:17 2000  Hiroshi IGARASHI  <iga@ruby-lang.org>

	* Makefile.tail (clean): pass value of variable DLLIBS.

Thu Oct 12 14:50:34 2000  Hiroshi IGARASHI  <iga@ruby-lang.org>

	* runtime/trace.c (get_mte): should be independent of dynamic location
	of functions.

Mon Jul 31 05:04:10 2000  Hiroshi IGARASHI  <iga@ruby-lang.org>

	* Configure, Makefile.tail, compiler/Makefile, disttest/Makefile,
	runtime/Makefile.head, runtime/Makefile.tail, shmtest/Makefile,
	test/Makefile: add shared library support(Linux/Solaris only).

Sat Mar 27 11:14:18 1999  Sekita Daigo  <sekita@mri.co.jp>

	* macro.kl1: fix the bug in replaceClase for the :- Dec form clause.

	* Configure: Added processes around shared libraries.

Sat Mar 27 11:10:19 1999  Sekita Daigo  <sekita@mri.co.jp>

	* struct.h: Change the macro definitions. (char *) -> (unsigned int)
	
1998-08-04  Sekita Daigo  <sekita@mri.co.jp>

	* Configure: remove const declaraion from the testing of sys_errlist.

1998-07-31  Sekita Daigo  <sekita@mri.co.jp>

	* ``polling mechanism'' may be used now in the PVM version. This
 	version of KLIC wait the message by polling in spite of the
 	normally by the interval timer. This mechanism is introduced
 	automatically for the platforms which does not equip setitmer()

klic-3.013/documents/workshop/tgif/tgif-QPL-4.2.5/tgif.c
/*
 * Author:      William Chia-Wei Cheng (bill.cheng@acm.org)
 *
 * Copyright (C) 2001-2009, William Chia-Wei Cheng.
 *
 * This file may be distributed under the terms of the Q Public License
 * as defined by Trolltech AS of Norway and appearing in the file
 * LICENSE.QPL included in the packaging of this file.
 *
 * THIS FILE IS PROVIDED AS IS WITH NO WARRANTY OF ANY KIND, INCLUDING
 * THE WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL,
 * INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
 * FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * @(#)$Header: /mm2/home/cvs/bc-src/tgif/tgif.c,v 1.17 2011/05/16 16:22:00 william Exp $
 */

#define _INCLUDE_FROM_TGIF_C_

#include "tgifdefs.h"

#include "color.e"
#include "dialog.e"
#include "exec.e"
#include "file.e"
#include "font.e"
#include "grid.e"
#include "mainloop.e"
#include "msg.e"
#include "names.e"
#include "obj.e"
#include "page.e"
#include "raster.e"
#include "setup.e"
#include "strtbl.e"
#include "tangram2.e"
#include "util.e"
#include "version.e"

/*
 * extern int	malloc_debug ARGS_DECL((int));
 */

int	lastFile=TRUE;

short pDrawFontAsc[] = {
    8, 10, 12, 14, 17, 22, 
    8, 10, 12, 14, 17, 22, 
    8, 10, 12, 14, 17, 23, 
    8, 10, 12, 14, 17, 22, 
    8,  9, 11, 13, 15, 19, 
    7,  9, 11, 12, 15, 20, 
    7,  9, 11, 13, 15, 19, 
    7,  9, 11, 12, 15, 20, 
    9, 11, 12, 14, 18, 24, 
    9, 11, 12, 14, 17, 24, 
    9, 11, 12, 14, 17, 24, 
    9, 11, 12, 14, 17, 24, 
    8, 11, 12, 14, 18, 23, 
    8, 11, 12, 15, 18, 24, 
    8, 11, 12, 14, 16, 23, 
    8, 11, 12, 14, 16, 24, 
    8, 10, 12, 14, 18, 24, 
    9, 14, 15, 18, 23, 30, 
   10, 14, 16, 17, 23, 30, 
    9, 13, 15, 17, 22, 30, 
    9, 13, 15, 18, 22, 30, 
    9, 12, 14, 15, 19, 26, 
    9, 12, 14, 15, 21, 26, 
    9, 12, 14, 15, 19, 26, 
    9, 12, 14, 15, 20, 26, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 18, 24, 31, 
   11, 14, 16, 19, 24, 32, 
   11, 15, 16, 19, 24, 33, 
   11, 14, 16, 18, 23, 32, 
   11, 15, 16, 19, 24, 32, 
   11, 12, 13, 14, 19, 27 
};
short pDrawFontDes[] = {
    2,  3,  3,  4,  4,  6, 
    2,  3,  3,  4,  4,  6, 
    2,  3,  3,  4,  5,  6, 
    2,  3,  3,  3,  4,  6, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  4,  5,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  4,  5,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  3,  5,  5, 
    2,  2,  3,  3,  5,  5, 
    2,  2,  3,  3,  5,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  3,  4,  5, 
    2,  2,  3,  3,  6,  5, 
    2,  2,  3,  3,  6,  5, 
    3,  4,  4,  6,  7,  8, 
    3,  3,  4,  4,  6,  7, 
    3,  3,  4,  4,  6,  7, 
    3,  4,  4,  5,  6,  7, 
    3,  3,  4,  4,  6,  7, 
    2,  3,  3,  4,  5,  6, 
    2,  4,  3,  5,  5,  7, 
    2,  3,  4,  4,  5,  6, 
    2,  4,  4,  5,  5,  7, 
    2,  3,  4,  4,  5,  7, 
    2,  3,  4,  5,  5,  7, 
    2,  3,  4,  5,  5,  7, 
    2,  3,  4,  5,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    2,  3,  3,  4,  5,  7, 
    4,  3,  4,  5,  5,  7 
};

static
void SetProgramName(FileName)
   char *FileName;
{
   char *c_ptr=UtilStrRChr(FileName, DIR_SEP);

   if (c_ptr != NULL && *(++c_ptr) != '\0') {
      UtilStrCpyN(progName, sizeof(progName), c_ptr);
   } else {
      UtilStrCpyN(progName, sizeof(progName), TOOL_NAME);
      UtilStrLower(progName);
   }
}

static
int PrTgifLoad(FileName)
   char *FileName;
{
   struct ObjRec *obj_ptr=NULL;
   char full_name[MAXPATHLENGTH+1], gzipped_fname[MAXPATHLENGTH+1];
   char tmp_filename[MAXPATHLENGTH+1], tmp_filefullpath[MAXPATHLENGTH+1];
   int read_status, obj_file=TRUE;
   FILE *fp=NULL;
   int tmp_linenum=0, gzipped=FALSE;

   if (strcmp(FileName, "-") == 0) {
      return TRUE;
   }
   *gzipped_fname = '\0';
   if (FileNameHasExtension(FileName, OBJ_FILE_TYPE, &gzipped, NULL)) {
      if (gzipped) {
         if (cmdLineOpenDisplay) {
            char *tmp_fname=NULL;

            if ((tmp_fname=GunzipFileIntoTemp(FileName)) == NULL) {
               return FALSE;
            } else {
               strcpy(gzipped_fname, FileName);
               strcpy(full_name, tmp_fname);
               free(tmp_fname);
            }
         } else {
            fprintf(stderr, TgLoadString(STID_TOOL_DONT_KNOW_HOW_RUN_GUNZIP),
                  TOOL_NAME);
            fprintf(stderr, "\n");
            fprintf(stderr, "%s\n",
                  TgLoadString(STID_SPECIFY_DISPLAY_ON_CMD_LINE));
            return FALSE;
         }
      } else {
         strcpy(full_name, FileName);
      }
   } else if (FileNameHasExtension(FileName, SYM_FILE_TYPE, NULL, NULL)) {
      strcpy(full_name, FileName);
      obj_file = FALSE;
   } else if (FileNameHasExtension(FileName, PIN_FILE_TYPE, NULL, NULL)) {
      strcpy(full_name, FileName);
      obj_file = FALSE;
   } else {
      sprintf(full_name, "%s.%s", FileName, OBJ_FILE_EXT);
   }
   if ((fp=fopen(full_name, "r")) == NULL) {
      fprintf(stderr, TgLoadString(STID_CANNOT_OPEN_FILE_FOR_READING),
            full_name);
      fprintf(stderr, "\n");
      if (*gzipped_fname != '\0') unlink(full_name);
      return FALSE;
   }

   strcpy(tmp_filefullpath, scanFileFullPath);
   strcpy(tmp_filename, scanFileName);
   tmp_linenum = scanLineNum;
   UtilStrCpyN(scanFileFullPath, sizeof(scanFileFullPath), full_name);
   strcpy(scanFileName, full_name);
   scanLineNum = 0;

   if (!(PRTGIF && cmdLineStdOut)) {
      if (!cmdLineQuiet) {
         fprintf(stderr, TgLoadCachedString(CSTID_READING_FILE_DOTS),
               full_name);
         fprintf(stderr, "\n");
      }
   }
   readingPageNum = loadedCurPageNum = 0;
   foundGoodStateObject = FALSE;
   while ((read_status=ReadObj(fp, &obj_ptr)) == TRUE) {
      if (obj_ptr != NULL) {
         AdjForOldVersion(obj_ptr);
         AddObj(NULL, topObj, obj_ptr);
      }
   }

   strcpy(scanFileFullPath, tmp_filefullpath);
   strcpy(scanFileName, tmp_filename);
   scanLineNum = tmp_linenum;

   fclose(fp);

   if (read_status == INVALID) {
      fprintf(stderr, TgLoadString(STID_FILE_VER_ABORT_TOOL),
            fileVersion, TOOL_NAME, TOOL_NAME, homePageURL);
      if (*gzipped_fname != '\0') unlink(full_name);
      return FALSE;
   }
   if (*gzipped_fname != '\0') {
      unlink(full_name);
      strcpy(full_name, gzipped_fname);
   }
   if (cmdLineOpenDisplay || *cmdLineGenParser != '\0') {
      char file_path[MAXPATHLENGTH+1];

      if (*full_name == DIR_SEP) {
         strcpy(file_path, full_name);
      } else {
         sprintf(file_path, "%s%c%s", curDir, DIR_SEP, full_name);
      }
      curFileDefined = TRUE;
      if (obj_file) {
         SetCurDir(file_path);
         *curSymDir = '\0';
      } else {
         SetCurSymDir(file_path);
      }
   }
   if (cmdLineHasPageNum) {
      if (cmdLinePageNum > lastPageNum) {
         if (lastPageNum == 1) {
            fprintf(stderr, TgLoadString(STID_FILE_SKIP_CONTAINS_ONE_PAGE),
                  FileName);
         } else {
            fprintf(stderr, TgLoadString(STID_FILE_SKIP_CONTAINS_NUM_PAGE),
                  FileName, lastPageNum);
         }
         fprintf(stderr, "\n");
         return FALSE;
      }
      GotoPageNum(cmdLinePageNum);
   }
   if (cmdLineColor && colorMenuItems == NULL) {
      fprintf(stderr, TgLoadString(STID_FILE_NOT_CONTAIN_COLOR_ABORT),
            TOOL_NAME);
      fprintf(stderr, "\n");
      return FALSE;
   }
   if (loadedCurPageNum > 0 && curPage != NULL && lastPageNum > 1) {
      /*
       * Well, it should go to loadedCurPageNum because it was the page
       * at which the file was saved.  But we go to lastPageNum to stay
       * compatible with Tangram-II since it was expected for tgif to
       * behave this way.
       */
      GotoPageNum(lastPageNum);
   }
   return TRUE;
}

static
void PrTgifDump(inbuf, tile_page_err_msg_id)
   char *inbuf;
   int tile_page_err_msg_id;
{
   if (inbuf != NULL && strcmp(inbuf, "-") != 0) {
      UtilStrCpyN(cmdLineOpenFile, sizeof(cmdLineOpenFile), inbuf);
   }
   if (cmdLineNoShowPageInEPS) showPageInEPS = FALSE;
   if (*cmdLineFileToExec != '\0') {
      if (!cmdLineQuiet) {
         fprintf(stderr, TgLoadString(STID_EXECUTING_GIVEN_SCRIPT),
               cmdLineFileToExec);
         fprintf(stderr, "\n");
      }
      StartExecCmdsFromFile();
   } else {
      colorDump = cmdLineColor;
      if (cmdLineOpenDisplay) {
         whereToPrint = cmdLineWhereToPrint;
      }
      if (cmdLineOneFilePerPage) {
         if (pageLayoutMode == PAGE_TILE) {
            fprintf(stderr, "%s\n", TgLoadString(tile_page_err_msg_id));
         } else {
            if (cmdLineHasPageNum) {
               sprintf(cmdLinePageNumStr, "%1d", cmdLinePageNum);
               GotoPageNum(cmdLinePageNum);
               if (inbuf != NULL) Dump(inbuf);
            } else {
               cmdLineHasPageNum = TRUE;
               for (cmdLinePageNum=1; cmdLinePageNum <= lastPageNum;
                     cmdLinePageNum++) {
                  sprintf(cmdLinePageNumStr, "%1d", cmdLinePageNum);
                  GotoPageNum(cmdLinePageNum);
                  if (inbuf != NULL) Dump(inbuf);
               }
               cmdLineHasPageNum = FALSE;
            }
         }
      } else {
         if (inbuf != NULL) Dump(inbuf);
      }
   }
   *cmdLineOpenFile = '\0';
   DelAllObj();
   if (cmdLineColor) CleanUpColors();
   if (cmdLineOpenDisplay) {
      CleanUp();
   }
}

static
void DoPrTgif(argc, argv, from_prtgif)
   int argc, from_prtgif;
   char *argv[];
{
   char inbuf[MAXSTRING+1];
   int len, argc_to_be;

   *scanFileFullPath = *scanFileName = '\0';
   scanLineNum = 0;

   if ((argc_to_be=ProcessPrTgifOptions(argc, argv, from_prtgif)) == INVALID) {
      return;
   }
   while (argc > argc_to_be) {
      argc--; argv++;
   }
   InitPaperSize();
   InitTmpDir();
   InitPatFill();
   if (PRTGIF && !cmdLineOpenDisplay) {
      InitDontReencode(NULL);
      InitPSFontCharSubs();
   }
   if (*cmdLineGenParser != '\0') {
      if (cmdLineTgrm2) {
         InitNames();
      }
   }
   if (argc <= 0) {
      if (cmdLineStdOut) {
         fprintf(stderr, "\n%s\n",
               TgLoadString(STID_SPEC_FILE_WHEN_STDOUT_USED));
         return;
      } else if (strcmp(cmdLineFileToExec, "-") == 0) {
         if (cmdLineOpenDisplay) {
            JustInit(NULL, NULL);
            Init2PatFill();
         }
         PrTgifDump(NULL, STID_CANT_ONEFPPAGE_IN_TILED_PAGE);
      } else {
         sprintf(gszMsgBox, TgLoadString(STID_TOOL_FILE_NAME_TO_PRINT),
               TOOL_NAME);
         fprintf(stderr, "\n%s> ", gszMsgBox);
         fflush(stderr);
         while (fgets(inbuf, MAXSTRING, stdin) != NULL) {
            len = strlen(inbuf);
            if (len > 0) {
               if (inbuf[--len] == '\n') inbuf[len] = '\0';
               if (*cmdLineGenParser != '\0') {
                  if (cmdLineTgrm2) {
                     if (PrTgifLoad(inbuf)) {
                        Tangram2GenerateChain(cmdLineGenParser);
                        DelAllObj();
                     } else {
                        fprintf(stderr, "\n");
                     }
                  }
               } else if (cmdLineDumpURL) {
                  UrlDump(inbuf);
               } else if (cmdLineDosEpsFilter) {
                  UtilStrCpyN(cmdLineOpenFile, sizeof(cmdLineOpenFile), inbuf);
                  DosEpsFilter(inbuf);
               } else {
                  if (cmdLineOpenDisplay) {
                     JustInit(NULL, NULL);
                  }
                  if (PrTgifLoad(inbuf)) {
                     PrTgifDump(inbuf, STID_CANT_ONEFPPAGE_IN_TILED_PAGE);
                  } else {
                     fprintf(stderr, "\n");
                  }
               }
            }
            sprintf(gszMsgBox, TgLoadString(STID_TOOL_FILE_NAME_TO_PRINT),
                  TOOL_NAME);
            fprintf(stderr, "\n%s> ", gszMsgBox);
            fflush(stderr);
         }
         fprintf(stderr, "\n");
         fflush(stderr);
      }
   } else {
      if (cmdLineStdOut && argc > 2) {
         fprintf(stderr, "\n%s\n",
               TgLoadString(STID_ONE_FILE_ONLY_WITH_STDOUT));
         return;
      }
      for ( ; argc > 0; argc--, argv++) {
         lastFile = (argc == 1);
         if (*cmdLineGenParser != '\0') {
            if (cmdLineTgrm2) {
               if (PrTgifLoad(*argv)) {
                  Tangram2GenerateChain(cmdLineGenParser);
                  DelAllObj();
               } else {
                  fprintf(stderr, "\n");
               }
            }
         } else if (cmdLineDumpURL) {
            UrlDump(*argv);
            if (argc > 1) {
               fprintf(stderr, "%s\n",
                     TgLoadString(STID_ONE_FILE_ONLY_RAW_PLUS_H));
            }
            break;
         } else if (cmdLineDosEpsFilter) {
            UtilStrCpyN(cmdLineOpenFile, sizeof(cmdLineOpenFile), *argv);
            DosEpsFilter(*argv);
            if (argc > 1) {
               fprintf(stderr, "%s\n",
                     TgLoadString(STID_ONE_FILE_ONLY_DOSEPSFILTER));
            }
            break;
         } else {
            if (cmdLineOpenDisplay) {
               JustInit(NULL, NULL);
            }
            if (PrTgifLoad(*argv)) {
               PrTgifDump(*argv, STID_CANT_PRTGIF_OFPP_TILED_PAGE);
            } else {
               fprintf(stderr, "\n");
            }
         }
      }
   }
   if (PRTGIF && !cmdLineOpenDisplay) CleanUpFonts();
   CleanUpPage();
   CleanUpResiduals();
   CleanUpPatFill();
   CleanUpStrTable();
}

/*
 * static
 * void Prompt2 (PromptStr, OpName, FileName)
 *    char	* PromptStr, * OpName, * FileName;
 * {
 *    char	inbuf[80];
 * 
 *    printf (PromptStr);
 *    fgets (inbuf, 80, stdin);
 *    sscanf (inbuf, "%s%s", OpName, FileName);
 * }
 *
 * static
 * void Prompt3 (PromptStr, AttrName, ColorName, ValName)
 *    char	* PromptStr, * AttrName, * ColorName, * ValName;
 * {
 *    char	inbuf[80];
 * 
 *    printf (PromptStr);
 *    fgets (inbuf, 80, stdin);
 *    sscanf (inbuf, "%s%s%s", AttrName, ColorName, ValName);
 * }
 */

static
void DoTgif(argc, argv)
   int argc;
   char *argv[];
   /* All these strangeness with strings are related to */
   /*    Prolog's foreign function interface. */
{
/*
 * register int i;
 */
   char op_name[80], file_name[(MAXPATHLENGTH<<1)+1];
   char *sp[6], *func_strp;
/*
 * char	color_name[80], val_name[80];
 * char	attr_name[80], speed_name[80], id_name[80];
 */

   if (!ProcessTgifOptions(argc, argv, file_name, sizeof(file_name))) return;

   if (file_name[0] == '\0') {
      /* do not translate -- program constants */
      MainLoop("init", "", &func_strp,
            &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
   } else {
      /* do not translate -- program constants */
      MainLoop("init", file_name, &func_strp,
            &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
   }
/*
 * for (i = 0; i < 6; i++)
 *    if (strcmp (sp[i], "") != 0)
 *       printf ("%s ", sp[i]);
 *    else
 *       break;
 * printf ("\n");
 */

   while (TRUE) {
      char s[80];

      strcpy(s, func_strp);
      s[4] = '\0';

      DeallocStrings(&func_strp,&sp[0],&sp[1],&sp[2],&sp[3],&sp[4],&sp[5]);

      /* do not translate -- program constants */
      if (strcmp(s, "Quit") == 0) {
         *file_name = '\0';
         MainLoop("quit", file_name, &func_strp,
               &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
         DeallocStrings(&func_strp,&sp[0],&sp[1],&sp[2],&sp[3],&sp[4],&sp[5]);
         break;
      }

      Msg(TgLoadCachedString(CSTID_RETURNED_FROM_BASIC_DRIVER));
/*
 *    Prompt2("Input an operation and a sub command.\n",op_name,file_name);
 *
 *    if (strcmp(op_name, "animate") == 0)
 *    {
 *       Prompt3("Input poly_id, speed, color.\n", id_name, speed_name,
 *             color_name);
 *       Animate(file_name, id_name, speed_name, color_name, &func_strp);
 *       printf("Animate RETURNs --> %s %s %s\n", func_strp, sp[0], sp[1]);
 *    }
 *    if (strcmp(op_name, "upd_attr_val") == 0)
 *    {
 *       Prompt3("Input attrname, color and value.\n", attr_name, color_name,
 *             val_name);
 *       UpdAttrVal(file_name, attr_name, color_name, val_name, &func_strp);
 *       printf("UpdAttrVal RETURNs --> %s %s %s\n", func_strp, sp[0], sp[1]);
 *    }
 */
      *op_name = *file_name = '\0';
      MainLoop(op_name, file_name, &func_strp,
            &sp[0], &sp[1], &sp[2], &sp[3], &sp[4], &sp[5]);
/*
 *    printf("RETURN --> %s ", func_strp);
 *    for (i = 0; i < 6; i++)
 *       if (strcmp(sp[i], "") != 0)
 *          printf("%s ", sp[i]);
 *       else
 *          break;
 *    printf("\n");
 */
   }
}

int main(argc, argv)
   int argc;
   char *argv[];
{
   register int i, from_prtgif;

   origArgC = argc;
   origArgV = argv;

   PRTGIF = FALSE;
   from_prtgif = FALSE;
   SetProgramName(*argv);
   for (i = 1; i < argc; i++) {
      /* do not translate -- program constants */
      if (strcmp(argv[i], "-prtgif")==0) {
         PRTGIF = TRUE;
         from_prtgif = TRUE;
         break;
      } else if ((strcmp(argv[i], "-print")==0) ||
            (strcmp(argv[i], "-PRINT")==0)) {
         PRTGIF = TRUE;
         break;
      }
   }
/*
 * malloc_debug (1);
 */
   if (PRTGIF) {
      DoPrTgif(argc, argv, from_prtgif);
   } else {
      DoTgif(argc, argv);
   }
   return 0;
}
